src/main/java/org/jetbrains/research/refactorinsight/ui/DiffHintLineMarkerProvider.java

```diff
@@ -0,0 +1,222 @@
+import com.intellij.openapi.ui.popup.JBPopup;
+import com.intellij.openapi.ui.popup.JBPopupFactory;
+import com.intellij.openapi.vcs.VcsException;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.FileViewProvider;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiIdentifier;
+import com.intellij.psi.impl.source.tree.LeafElement;
+import com.intellij.psi.util.PsiUtilCore;
+import com.intellij.ui.awt.RelativePoint;
+import com.intellij.util.Function;
+import com.intellij.util.ui.JBEmptyBorder;
+import git4idea.GitCommit;
+import git4idea.history.GitHistoryUtils;
+import icons.RefactorInsightIcons;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.research.refactorinsight.RefactorInsightBundle;
+import org.jetbrains.research.refactorinsight.data.RefactoringEntry;
+import org.jetbrains.research.refactorinsight.data.RefactoringInfo;
+import org.jetbrains.research.refactorinsight.services.MiningService;
+import org.jetbrains.research.refactorinsight.ui.windows.DiffWindow;
+
+import javax.swing.*;
+import java.awt.event.MouseEvent;
+import java.util.*;
+
+public class DiffHintLineMarkerProvider extends LineMarkerProviderDescriptor {
+    private static final String DIFF_WINDOW_CLASS_NAME_PREFIX = DiffContentFactoryImpl.class.getName() + "$";
+
+    @Override
+    public @Nullable("null means disabled")
+    @GutterName String getName() {
+        return "RefactorInsight gutter";
+    }
+
+    @Override
+    public @Nullable Icon getIcon() {
+        return RefactorInsightIcons.toggle;
+    }
+
+    @Override
+    public LineMarkerInfo<?> getLineMarkerInfo(@NotNull PsiElement element) {
+        return null;
+    }
+
+    @Override
+    public void collectSlowLineMarkers(@NotNull List<? extends PsiElement> elements,
+                                       @NotNull Collection<? super LineMarkerInfo<?>> result) {
+        if (elements.isEmpty()) return;
+        List<RefactoringInfo> refactoringInfos = getRefactoringInfos(elements.get(0));
+        if (refactoringInfos == null) return;
+
+        boolean isRight = isRightPartOfDiff(elements.get(0));
+        Map<Integer, Set<RefactoringInfo>> refactoringsMap = new HashMap<>();
+
+        for (PsiElement element : elements) {
+            if (!isIdentifier(element)) continue;
+
+            int lineNumber = getLineNumber(element);
+            int textOffset = element.getTextOffset();
+            for (RefactoringInfo refactoringInfo : refactoringInfos) {
+                refactoringsMap.putIfAbsent(lineNumber, new HashSet<>());
+                if (refactoringInfo.containsElement(lineNumber, textOffset, isRight) &&
+                        !refactoringsMap.get(lineNumber).contains(refactoringInfo)) {
+                    RefactoringInfoHint info = new RefactoringInfoHint(element, e -> "Refactoring detected", refactoringInfo);
```

Comment:
Please, move message to `RefactorInsightBundle.properties` (`refactoring.detected.hint = Refactoring detected`) and use `RefactorInsightBundle.message("refactoring.detected.hint")` to get it here.